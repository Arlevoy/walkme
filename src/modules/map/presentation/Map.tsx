import * as React from 'react';
import MapView from 'react-native-maps';
import { StyleSheet, Text, View, Dimensions } from 'react-native';
import { theme } from '../../../shared/theme';

export const Map: React.FunctionComponent = () => {
  return (
    <View style={styles.container}>
      <MapView style={styles.map} />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: theme.colors.white,
    alignItems: 'center',
    justifyContent: 'center',
  },
  map: {
    width: Dimensions.get('window').width,
    height: Dimensions.get('window').height,
  },
});
