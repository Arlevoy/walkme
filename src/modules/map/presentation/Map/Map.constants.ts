export const INITIAL_REGION = {
  latitude: 48.85625516456069,
  latitudeDelta: 0.18850770649190451,
  longitude: 2.343346362242182,
  longitudeDelta: 0.1475828137637336,
};

export const FIRST_ROUTE = [
  {
    latitude: 48.86162558742782,
    latitudeDelta: 0.18850770649190451,
    longitude: 2.3808150116565927,
    longitudeDelta: 0.1475828137637336,
  },
  {
    latitude: 48.85829269782787,
    latitudeDelta: 0.18850770649190451,
    longitude: 2.3798557320696117,
    longitudeDelta: 0.1475828137637336,
  },
  {
    latitude: 48.862197269529446,
    latitudeDelta: 0.18850770649190451,
    longitude: 2.394183622649873,
    longitudeDelta: 0.1475828137637336,
  },
];

export const HOME = {
  latitude: 48.85937441095902,
  latitudeDelta: 0.005,
  longitude: 2.3898020245080662,
  longitudeDelta: 0.005,
};
