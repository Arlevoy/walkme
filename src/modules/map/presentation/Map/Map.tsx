import React, { FunctionComponent, useRef, useState } from 'react';

import {
  Dimensions,
  SafeAreaView,
  StyleSheet,
  Text,
  TextStyle,
  TouchableOpacity,
  View,
  ViewStyle,
} from 'react-native';
import MapView, { Marker } from 'react-native-maps';
import { theme } from '../../../../shared/theme';

import { useGetCurrentRegion } from '#modules/map/infra/useGetCurrentPosition.hook';
import { FIRST_ROUTE } from '#modules/map/presentation/Map/Map.constants';
import { OverlayCardInterface } from '#modules/map/presentation/OverlayCard/OverlayCard.interface';
import { CurrentPositionIcon } from '#shared/assets/icons/CurrentPositionIcon';
import { OverlayCard } from '../OverlayCard/OverlayCard';

interface MapStyles {
  container: ViewStyle;
  map: ViewStyle;
  currentPositionIcon: ViewStyle;
  startRouteButton: ViewStyle;
  startRouteText: TextStyle;
}

const styles = StyleSheet.create<MapStyles>({
  container: {
    alignItems: 'center',
    backgroundColor: theme.colors.white,
    flex: 1,
    justifyContent: 'center',
  },
  currentPositionIcon: {
    padding: theme.spacing.unit * 3,
    position: 'absolute',
    right: 0,
    top: 50,
  },
  map: {
    height: Dimensions.get('window').height,
    width: Dimensions.get('window').width,
  },
  startRouteButton: {
    backgroundColor: theme.colors.white,
    borderRadius: theme.spacing.unit,
    opacity: 0.9,
    padding: theme.spacing.unit * 2,
    position: 'absolute',
    top: 0,
  },
  startRouteText: {
    color: theme.colors.primary,
    fontSize: theme.font.size.L,
  },
});

const HOME = {
  latitude: 48.85937441095902,
  latitudeDelta: 0.005,
  longitude: 2.489802024508066,
  longitudeDelta: 0.005,
};

const DEFAULT_PADDING = {
  top: theme.spacing.unit * 4,
  bottom: theme.spacing.unit * 4,
  right: theme.spacing.unit * 4,
  left: theme.spacing.unit * 4,
};

export const Map: FunctionComponent = () => {
  const [overlayCardStatus, setOverlayCardStatus] =
    useState<OverlayCardInterface['status']>('HIDDEN');
  const onMarkerPress = () => {
    setOverlayCardStatus('DISPLAYED');
  };
  const mapViewRef = useRef<MapView>(null);
  const { currentRegion, onCurrentPositionPress } = useGetCurrentRegion(mapViewRef);

  const onCloseOverlayCard = () => setOverlayCardStatus('HIDDEN');

  const onStartRoutePress = () => {
    if (!mapViewRef || !mapViewRef.current) return;
    mapViewRef.current.fitToCoordinates(FIRST_ROUTE, { edgePadding: DEFAULT_PADDING });
  };

  return (
    <SafeAreaView>
      <View testID="map" style={styles.container}>
        <MapView
          ref={mapViewRef}
          style={styles.map}
          initialRegion={currentRegion}
          showsMyLocationButton={true}
          showsUserLocation={true}
        >
          <Marker key={1} coordinate={HOME} onPress={onMarkerPress} testID={'marker'} />
          {FIRST_ROUTE.map((region) => (
            <Marker key={region.latitude} coordinate={region} onPress={onMarkerPress} />
          ))}
        </MapView>
        <TouchableOpacity onPress={onCurrentPositionPress} style={styles.currentPositionIcon}>
          <CurrentPositionIcon />
        </TouchableOpacity>
        <TouchableOpacity onPress={onStartRoutePress} style={styles.startRouteButton}>
          <Text style={styles.startRouteText}>DÃ©marre ton premier parcours !</Text>
        </TouchableOpacity>
        <OverlayCard status={overlayCardStatus} onClose={onCloseOverlayCard} />
      </View>
    </SafeAreaView>
  );
};
