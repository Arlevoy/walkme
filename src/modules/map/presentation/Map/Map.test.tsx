import { Map } from '#modules/map/presentation/Map/Map';
import { fireEvent, render } from '@testing-library/react-native';
import React from 'react';

describe('Map', () => {
  it('should render Map', () => {
    const screen = render(<Map />);

    expect(screen).toMatchSnapshot();
  });

  it('should display card on marker presss', async () => {
    const screen = render(<Map />);

    const marker = screen.getByTestId('marker');

    fireEvent.press(marker);

    expect(screen.getByText('My Home')).toBeTruthy();
  });
});
