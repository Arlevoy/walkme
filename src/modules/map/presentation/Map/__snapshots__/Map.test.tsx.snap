// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Map should render Map 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#FFFFFF",
      "flex": 1,
      "justifyContent": "center",
    }
  }
  testID="map"
>
  <AIRMap
    initialRegion={null}
    onChange={[Function]}
    onLayout={[Function]}
    onMapReady={[Function]}
    onMarkerPress={[Function]}
    region={
      Object {
        "latitude": 48.85937441095902,
        "latitudeDelta": 0.005,
        "longitude": 2.3898020245080662,
        "longitudeDelta": 0.005,
      }
    }
    style={
      Object {
        "height": 1334,
        "width": 750,
      }
    }
  >
    <AIRMapMarker
      coordinate={
        Object {
          "latitude": 48.85937441095902,
          "latitudeDelta": 0.005,
          "longitude": 2.3898020245080662,
          "longitudeDelta": 0.005,
        }
      }
      onPress={[Function]}
      stopPropagation={false}
      style={
        Array [
          Object {
            "backgroundColor": "transparent",
            "position": "absolute",
          },
          undefined,
        ]
      }
      testID="marker"
    />
  </AIRMap>
  <Modal
    animationType="none"
    deviceHeight={null}
    deviceWidth={null}
    hardwareAccelerated={false}
    hideModalContentWhileAnimating={false}
    onBackdropPress={[Function]}
    onModalHide={[Function]}
    onModalWillHide={[Function]}
    onModalWillShow={[Function]}
    onRequestClose={[Function]}
    panResponderThreshold={4}
    scrollHorizontal={false}
    scrollOffset={0}
    scrollOffsetMax={0}
    scrollTo={null}
    statusBarTranslucent={false}
    supportedOrientations={
      Array [
        "portrait",
        "landscape",
      ]
    }
    swipeThreshold={100}
    transparent={true}
    visible={false}
  />
</View>
`;
